### PR TITLE
KOGITO-406: [DMN Designer] Unit tests: kie-wb-common-dmn-webapp-kogito-marshaller

### PR DESCRIPTION
**Pull request by manstis**
See https://issues.redhat.com/browse/KOGITO-406

This PR contains the basic setup for performing the tests using the client-side marshaller.

It contains a couple of basic tests.

A follow up JIRA will be created (probably many!) to track addition of more tests..

### CODE REVIEW

**Comment by manstis** [comment]
As @gitgabrio pointed out the tests in this PR are really unit tests and not integration tests...

At present consider them a place-holder for _real_ integration tests that will load XML, convert it into Stunner's model and then back to XML. A comparison of the XML string will hopefully be possible. 

The main point of this PR is to add the integration test project and scaffolding for more tests.

**Comment by manstis** [comment]
:-( I'll take a look a little later.

**Comment by manstis** [comment]
@karreiro Even after deleting my local `.m2` I could not replicate.

However I've added a new commit that I hope fixes things...

**Comment by manstis** [comment]
@jomarko @karreiro Could we please try to get this merged before the Christmas shutdown.

**Comment by manstis** [comment]
Thanks @karreiro wouldn't it be nice to have regular separate test classes!

However the bootstrap of the tests requires asynchronous loading of multiple JavaScript files and I kick off "tests" as the last step in the asynchronous process. If you have a bright idea how to only load the scripts once and have tests run after they've loaded I am open to change. Unfortunately GWTs "before" method is "before each test". Finally GWT only supports junit.framework.

What I have tried to do is to best emulate normal test definition, execution and reporting as best I can when setup of the tests requires completion of an asynchronous operation.

File: kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/pom.xml [hunk 1]
```diff
@@ -121,6 +121,10 @@
       <artifactId>kie-wb-common-dmn-webapp-kogito-marshaller</artifactId>
       <scope>provided</scope>
     </dependency>
+    <dependency>
+      <groupId>org.uberfire</groupId>
+      <artifactId>uberfire-wires-core-grids</artifactId>
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/pom.xml]: mvn says unused declared dependency

File: kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/pom.xml [hunk 2]
```diff
@@ -259,6 +263,98 @@
       <scope>test</scope>
     </dependency>
 
+    <!-- Source dependencies required for integration tests -->
+    <dependency>
+      <groupId>org.kie.soup</groupId>
+      <artifactId>kie-soup-commons</artifactId>
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/pom.xml]: mvn says it is unused

File: kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/pom.xml [hunk 3]
```diff
@@ -259,6 +263,98 @@
...
+      <classifier>sources</classifier>
+    </dependency>
+    <dependency>
+      <groupId>org.uberfire</groupId>
+      <artifactId>uberfire-api</artifactId>
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/pom.xml]: mvn says it is unused

File: kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/pom.xml [hunk 4]
```diff
@@ -259,6 +263,98 @@
...
+      <classifier>sources</classifier>
+    </dependency>
+    <dependency>
+      <groupId>org.uberfire</groupId>
+      <artifactId>uberfire-client-api</artifactId>
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/pom.xml]: mvn says it is unused

File: kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/src/it/kie-wb-common-dmn-webapp-kogito-marshaller-tests/src/test/java/org/kie/workbench/common/dmn/webapp/kogito/marshaller/mapper/DMNMarshallerKogitoUnmarshallerTest.java [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+public class DMNMarshallerKogitoUnmarshallerTest extends IOCClientTestCase {
+
+    private static final Logger LOGGER = Logger.getLogger(DMNMarshallerKogitoUnmarshallerTest.class.getName());
+
+    @Override
+    public String getModuleName() {
+        return "org.kie.workbench.common.dmn.webapp.kogito.marshaller.DMNMarshallerKogitoUnmarshallerTest";
+    }
+
+    public void testUnmarshall() {
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/src/it/kie-wb-common-dmn-webapp-kogito-marshaller-tests/src/test/java/org/kie/workbench/common/dmn/webapp/kogito/marshaller/mapper/DMNMarshallerKogitoUnmarshallerTest.java]: @Test is not needed?

Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/src/it/kie-wb-common-dmn-webapp-kogito-marshaller-tests/src/test/java/org/kie/workbench/common/dmn/webapp/kogito/marshaller/mapper/DMNMarshallerKogitoUnmarshallerTest.java]: No. GWT only emulates `junit.framework` in the client/browser (where this code runs) and hence we need to use the _old_ way of prefixing test methods with `test`.

Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/src/it/kie-wb-common-dmn-webapp-kogito-marshaller-tests/src/test/java/org/kie/workbench/common/dmn/webapp/kogito/marshaller/mapper/DMNMarshallerKogitoUnmarshallerTest.java]: See http://www.gwtproject.org/doc/latest/DevGuideTesting.html

File: kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/src/it/kie-wb-common-dmn-webapp-kogito-marshaller-tests/src/test/java/org/kie/workbench/common/dmn/webapp/kogito/marshaller/mapper/DMNMarshallerKogitoUnmarshallerTest.java [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+import java.util.logging.Logger;
+
+import org.jboss.errai.ioc.client.IOCClientTestCase;
+import org.jboss.errai.ioc.client.container.IOC;
+import org.jboss.errai.ioc.client.container.SyncBeanDef;
+import org.kie.workbench.common.dmn.webapp.kogito.marshaller.mapper.tests.EmptyDiagramTest;
+import org.kie.workbench.common.dmn.webapp.kogito.marshaller.mapper.tests.InputDataTest;
+import org.kie.workbench.common.stunner.kogito.client.service.KogitoClientDiagramService;
+
+public class DMNMarshallerKogitoUnmarshallerTest extends IOCClientTestCase {
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/src/it/kie-wb-common-dmn-webapp-kogito-marshaller-tests/src/test/java/org/kie/workbench/common/dmn/webapp/kogito/marshaller/mapper/DMNMarshallerKogitoUnmarshallerTest.java]: Can't be the project strucutured in a way that I fetch changes from git(hub) restart my idea and it is loaded? It is hard to review when I can not navigate trough the code.

Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/src/it/kie-wb-common-dmn-webapp-kogito-marshaller-tests/src/test/java/org/kie/workbench/common/dmn/webapp/kogito/marshaller/mapper/DMNMarshallerKogitoUnmarshallerTest.java]: You can load the _test_ module into your IDE (that's how I wrote it).

Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/src/it/kie-wb-common-dmn-webapp-kogito-marshaller-tests/src/test/java/org/kie/workbench/common/dmn/webapp/kogito/marshaller/mapper/DMNMarshallerKogitoUnmarshallerTest.java]: Although I am in two minds to move this test module to the top level DMN folder (and not have it as an integration test). WDYT?

File: kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/src/it/kie-wb-common-dmn-webapp-kogito-marshaller-tests/src/test/java/org/kie/workbench/common/dmn/webapp/kogito/marshaller/mapper/DMNTest.java [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.kie.workbench.common.dmn.webapp.kogito.marshaller.mapper;
+
+import org.kie.workbench.common.stunner.core.diagram.Diagram;
+import org.kie.workbench.common.stunner.kogito.client.service.KogitoClientDiagramService;
+
+/**
+ * Definition of a DMN test.
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/src/it/kie-wb-common-dmn-webapp-kogito-marshaller-tests/src/test/java/org/kie/workbench/common/dmn/webapp/kogito/marshaller/mapper/DMNTest.java]: I think we should be more specific what DMNTest is. What are his inputs, what are his outputs (API basically).

File: kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/src/it/kie-wb-common-dmn-webapp-kogito-marshaller-tests/src/test/java/org/kie/workbench/common/dmn/webapp/kogito/marshaller/mapper/DMNUnmarshallerTestBootstrap.java [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+            this.success = success;
+        }
+
+        @Override
+        public void onFailure(final Exception reason) {
+            fail(reason.getMessage());
+        }
+
+        @Override
+        public void onSuccess(final Void result) {
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/src/it/kie-wb-common-dmn-webapp-kogito-marshaller-tests/src/test/java/org/kie/workbench/common/dmn/webapp/kogito/marshaller/mapper/DMNUnmarshallerTestBootstrap.java]: `Void` ?

File: kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/src/it/kie-wb-common-dmn-webapp-kogito-marshaller-tests/src/test/java/org/kie/workbench/common/dmn/webapp/kogito/marshaller/mapper/tests/BaseDMNTest.java [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+        } catch (Exception e) {
+            LOGGER.info(e.getMessage());
+        }
+    }
+
+    protected void logDiagramBasics(final Diagram diagram) {
+        final StringBuilder sb = new StringBuilder("\n");
+        sb.append("-> Name: " + diagram.getName()).append("\n");
+        sb.append("-> DefinitionSetId: " + diagram.getMetadata().getDefinitionSetId()).append("\n");
+        sb.append("-> Nodes").append("\n");
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/src/it/kie-wb-common-dmn-webapp-kogito-marshaller-tests/src/test/java/org/kie/workbench/common/dmn/webapp/kogito/marshaller/mapper/tests/BaseDMNTest.java]: Maybe worth to implement `toString` in Diagram?

File: kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/src/it/kie-wb-common-dmn-webapp-kogito-marshaller-tests/src/test/java/org/kie/workbench/common/dmn/webapp/kogito/marshaller/mapper/tests/InputDataTest.java [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+import org.kie.workbench.common.stunner.core.diagram.Diagram;
+import org.kie.workbench.common.stunner.core.graph.Node;
+import org.kie.workbench.common.stunner.core.graph.content.definition.Definition;
+import org.kie.workbench.common.stunner.core.graph.content.view.View;
+import org.kie.workbench.common.stunner.core.graph.util.GraphUtils;
+import org.kie.workbench.common.stunner.kogito.client.service.KogitoClientDiagramService;
+
+import static junit.framework.TestCase.assertEquals;
+import static junit.framework.TestCase.assertNotNull;
+import static junit.framework.TestCase.assertTrue;
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/src/it/kie-wb-common-dmn-webapp-kogito-marshaller-tests/src/test/java/org/kie/workbench/common/dmn/webapp/kogito/marshaller/mapper/tests/InputDataTest.java]: are packages correct?

File: kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/src/it/kie-wb-common-dmn-webapp-kogito-marshaller-tests/src/test/java/org/kie/workbench/common/dmn/webapp/kogito/marshaller/workaround/TestDirectoryUrlType.java [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+/**
+ * This prevents errai's code generation complaining that it does not have a {@link Vfs.UrlType} for
+ * file:/x/y/z/target/test-classes. See {@link Vfs.DefaultUrlTypes} directory that expects file:/x/y/z/target/classes
+ * to exist. This is not the case if there are no src/main/java classes in the module.
+ */
+public class TestDirectoryUrlType implements Vfs.UrlType {
+
+    @Override
+    public boolean matches(final URL url) {
+        return url.getProtocol().equals("file");
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/src/it/kie-wb-common-dmn-webapp-kogito-marshaller-tests/src/test/java/org/kie/workbench/common/dmn/webapp/kogito/marshaller/workaround/TestDirectoryUrlType.java]: objects.equals please

File: kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/src/it/kie-wb-common-dmn-webapp-kogito-marshaller-tests/src/test/java/org/kie/workbench/common/dmn/webapp/kogito/marshaller/mapper/tests/InputDataTest.java [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+
+    @Override
+    public String getTestName() {
+        return getClass().getName();
+    }
+
+    @Override
+    public void run(final KogitoClientDiagramService service) throws AssertionError {
+        test(service,
+             UnmarshallerXMLTests.INSTANCE.inputData().getText());
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/src/it/kie-wb-common-dmn-webapp-kogito-marshaller-tests/src/test/java/org/kie/workbench/common/dmn/webapp/kogito/marshaller/mapper/tests/InputDataTest.java]: I imported the diagram into the business central, the node has event not set the data type? Is that intentional?

Will we create a lot of non complete dmn files just for testing? From my point of view would be useful to use more complex dmn files wshi are valid (if we are not testing negative scenario). I think it could descrease total file of dmn files we maintain. would be easy to point field people to some getting started diagrams. Just Idea, I can be wrong for sure.

Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/src/it/kie-wb-common-dmn-webapp-kogito-marshaller-tests/src/test/java/org/kie/workbench/common/dmn/webapp/kogito/marshaller/mapper/tests/InputDataTest.java]: The _unit tests_ in this PR just demonstrate how to write a test using the new module. The plan is to remove them and make proper _integration tests_ that use _real_ DMN files and check their marshalling from XML->objects->XML. This is covered by https://issues.redhat.com/browse/KOGITO-778 (that will probably be broken into smaller tasks).

File: kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/src/it/kie-wb-common-dmn-webapp-kogito-marshaller-tests/src/test/java/org/kie/workbench/common/dmn/webapp/kogito/marshaller/workaround/TestDirectoryUrlType.java (hunk 11 above)
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/src/it/kie-wb-common-dmn-webapp-kogito-marshaller-tests/src/test/java/org/kie/workbench/common/dmn/webapp/kogito/marshaller/workaround/TestDirectoryUrlType.java]: Done.

File: kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/src/it/kie-wb-common-dmn-webapp-kogito-marshaller-tests/src/test/java/org/kie/workbench/common/dmn/webapp/kogito/marshaller/mapper/tests/InputDataTest.java (hunk 10 above)
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/src/it/kie-wb-common-dmn-webapp-kogito-marshaller-tests/src/test/java/org/kie/workbench/common/dmn/webapp/kogito/marshaller/mapper/tests/InputDataTest.java]: Yes unfortunately :-(

This whole new module, including tests, is compiled by GWT into JavaScript. GWT only emulates `junit.framework` and hence we're not able to use _regular_ JUnit nor AssertJ.

File: kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/src/it/kie-wb-common-dmn-webapp-kogito-marshaller-tests/src/test/java/org/kie/workbench/common/dmn/webapp/kogito/marshaller/mapper/tests/BaseDMNTest.java (hunk 9 above)
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/src/it/kie-wb-common-dmn-webapp-kogito-marshaller-tests/src/test/java/org/kie/workbench/common/dmn/webapp/kogito/marshaller/mapper/tests/BaseDMNTest.java]: Nice idea. See https://issues.redhat.com/browse/JBPM-8959

File: kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/src/it/kie-wb-common-dmn-webapp-kogito-marshaller-tests/src/test/java/org/kie/workbench/common/dmn/webapp/kogito/marshaller/mapper/DMNUnmarshallerTestBootstrap.java (hunk 8 above)
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/src/it/kie-wb-common-dmn-webapp-kogito-marshaller-tests/src/test/java/org/kie/workbench/common/dmn/webapp/kogito/marshaller/mapper/DMNUnmarshallerTestBootstrap.java]: Yes. The _callback_ for GWT's `ScriptInjector` is of type `Callback<Void, Exception>`. The `Callback` interface uses the first generic type as the success method parameter type.. I assume the reasoning being that the _success_ of injecting a script is meaningless (and hence void).

File: kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/pom.xml (hunk 4 above)
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/pom.xml]: It seems to be required by the integration test/invoker plugin for some reason. If I delete the dependency declaration then `kie-wb-common-dmn-webapp-kogito-marshaller-tests` complains about unresolved classes. However if I move the dependency declaration to `kie-wb-common-dmn-webapp-kogito-marshaller-tests` compilation still fails with an error stating the dependency cannot be resolved.

File: kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/pom.xml (hunk 3 above)
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/pom.xml]: It seems to be required by the integration test/invoker plugin for some reason. If I delete the dependency declaration then `kie-wb-common-dmn-webapp-kogito-marshaller-tests` complains about unresolved classes. However if I move the dependency declaration to `kie-wb-common-dmn-webapp-kogito-marshaller-tests` compilation still fails with an error stating the dependency cannot be resolved.

File: kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/pom.xml (hunk 2 above)
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/pom.xml]: It seems to be required by the integration test/invoker plugin for some reason. If I delete the dependency declaration then `kie-wb-common-dmn-webapp-kogito-marshaller-tests` complains about unresolved classes. However if I move the dependency declaration to `kie-wb-common-dmn-webapp-kogito-marshaller-tests` compilation still fails with an error stating the dependency cannot be resolved.

File: kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/pom.xml (hunk 1 above)
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/pom.xml]: Moved to `kie-wb-common-dmn-webapp-kogito-marshaller-tests`.

**Comment by manstis** [comment]
Jenkins please retest this.

**Comment by manstis** [comment]
Jenkins please retest this.

**Comment by manstis** [comment]
Jenkins please retest this.
